File: client/src/components/cards/maps.ts
/**
 * Map registry: the room a match is played in.
 *
 * A map is a scene, a table and an accent colour. It touches no rule and no
 * card. The *choice* is the server's (see server/game/maps.go), so every seat at
 * one table plays in one room; this file only says what each id looks like.
 *
 * Nothing here is a picture. The scene is rendered in the browser by the
 * isometric engine in `components/scene/` from the three ids the server deals
 * (`map_id`, `time_of_day`, `weather`), and the table is drawn in CSS by
 * `GameBoard.svelte` from the materials named below. A map used to be two
 * photographs and a rectangle measured off them; what replaced that is a place
 * that has an hour and a sky, and a table whose felt and rim belong to it.
 *
 * Deliberately not themed. The room does not follow light/dark like the rest of
 * the UI, for the same reason a card face does not: it is a place, not a
 * surface, and the same place in two colour schemes is two places.
 */
import { TIMES, WEATHERS, isTime, isWeather, type TimeOfDay, type Weather } from '../scene/sky'

export type MapId = 'neon' | 'rune' | 'velvet' | 'orbit' | 'sakura' | 'marina'

/**
 * What the rim and the edge are made of: a noble material, one per room.
 * `lacquer` is a piano finish (Neon's black, Sakura's urushi), `oak` an oiled
 * quartersawn oak, `burl` a varnished walnut burl, `titanium` brushed metal,
 * `teak` a yacht's varnished teak.
 */
export type RimKind = 'lacquer' | 'oak' | 'burl' | 'titanium' | 'teak'
/** The edge's profile, seen in section: what the render sweeps round the felt. */
export type EdgeProfile = 'bullnose' | 'ogee' | 'knife' | 'square'
/** What the table stands on. */
export type PedestalKind = 'tulip' | 'turned' | 'deco' | 'tripod' | 'legs' | 'capstan'

/**
 * What the table is made of, per room. Every colour is a CSS colour.
 *
 * `felt` and `feltDeep` are the playing surface's gradient, `rim` the material
 * of the edge, `rimLight` the sheen it catches, `base` the pedestal, `inlay`
 * the metal set into the wood (the bead round the felt and the play
 * direction's chevrons). The rest names the materials, so the CSS can draw
 * their grain (`tableSurface.ts`) and the render can sweep the edge and turn
 * the pedestal (`scene/maps/vista.ts: vistaTable`). The hour tints the sheen
 * (`--scene-tint`) and dims the whole object (`--scene-dark`); the materials
 * themselves never change with it, because a table is a physical thing and
 * night does not repaint it.
 */
export interface TableMaterials {
  felt: string
  feltDeep: string
  rim: string
  rimLight: string
  base: string
  inlay: string
  rimKind: RimKind
  /** The material's second colour: the wood's rings, the lacquer's depth, the brushing. */
  grain: string
  /** Metal flakes sown in the finish (Sakura's gold in the urushi, *nashiji*). */
  fleck?: string
  edge: EdgeProfile
  pedestal: PedestalKind
  /** The pedestal is stone and this is its vein (Velvet's black marble). */
  baseVein?: string
}

export interface MapDef {
  id: MapId
  table: TableMaterials
  /**
   * The map's colour, used for the light the room casts on the board: the glow
   * pooled under the table, the ambient wash at the edges, and the direction
   * ring's chevrons.
   *
   * It deliberately does NOT reach the brand: the "your turn" pill, the active
   * seat's gold and the card faces are the same in every room. Those are what a
   * viewer reads the game state off, and a state cue that changes colour with
   * the scenery is a cue that has to be re-learned six times.
   */
  accent: string
  /** A dimmer companion to `accent`, for the wide low-opacity washes. */
  accentDeep: string
  /**
   * Nothing falls from this sky: a storm here is light and dust, never rain.
   * The overlay reads it; the server's weather list is unchanged by it.
   */
  dry?: boolean
  /**
   * The skies this room can be dealt under. A mirror of `game.MapWeathers`,
   * pinned to it by `maps.test.ts`: the server draws from its list, and this one
   * says which the client can draw, so the two must agree.
   */
  weathers: readonly Weather[]
}

export const MAPS: Record<MapId, MapDef> = {
  neon: {
    id: 'neon',
    table: {
      felt: '#2a2046',
      feltDeep: '#0d0a1c',
      rim: '#0e0c14',
      rimLight: '#c9b8ff',
      base: '#b9bec9',
      inlay: '#dfe3ec',
      rimKind: 'lacquer',
      grain: '#2c2640',
      edge: 'knife',
      pedestal: 'tulip',
    },
    accent: '#c56bff',
    accentDeep: '#5a1e9c',
    weathers: ['clear', 'cloudy', 'rain', 'storm', 'snow', 'fog'],
  },
  rune: {
    id: 'rune',
    table: {
      felt: '#2b4431',
      feltDeep: '#0f1c13',
      rim: '#7a5130',
      rimLight: '#e0ad72',
      base: '#5a3719',
      inlay: '#c9955a',
      rimKind: 'oak',
      grain: '#3b2412',
      edge: 'square',
      pedestal: 'turned',
    },
    accent: '#ffab52',
    accentDeep: '#6d3410',
    weathers: ['clear', 'cloudy', 'rain', 'storm', 'snow', 'fog'],
  },
  velvet: {
    id: 'velvet',
    table: {
      felt: '#6e1729',
      feltDeep: '#2a0610',
      rim: '#6b3d1c',
      rimLight: '#f3cf94',
      base: '#141214',
      inlay: '#e9c46e',
      rimKind: 'burl',
      grain: '#2b1406',
      edge: 'ogee',
      pedestal: 'deco',
      baseVein: '#d9d2c6',
    },
    accent: '#f0c46a',
    accentDeep: '#5e3a12',
    weathers: ['clear', 'cloudy', 'rain', 'snow', 'fog'],
  },
  orbit: {
    id: 'orbit',
    table: {
      felt: '#133342',
      feltDeep: '#061520',
      rim: '#868e99',
      rimLight: '#eef4fa',
      base: '#7a828d',
      inlay: '#bfe8f5',
      rimKind: 'titanium',
      grain: '#5c646f',
      edge: 'knife',
      pedestal: 'tripod',
    },
    accent: '#4fd6ff',
    accentDeep: '#123a63',
    dry: true,
    weathers: ['clear', 'fog', 'storm'],
  },
  sakura: {
    id: 'sakura',
    table: {
      felt: '#27432f',
      feltDeep: '#0e1d14',
      rim: '#9a1f17',
      rimLight: '#ff9a7a',
      base: '#17110f',
      inlay: '#e6bc62',
      rimKind: 'lacquer',
      grain: '#4f0a07',
      fleck: '#e8c065',
      edge: 'bullnose',
      pedestal: 'legs',
    },
    accent: '#ff8fb8',
    accentDeep: '#7a2a4a',
    weathers: ['clear', 'cloudy', 'rain', 'snow', 'fog'],
  },
  marina: {
    id: 'marina',
    table: {
      felt: '#15344f',
      feltDeep: '#081726',
      rim: '#8f5d2e',
      rimLight: '#f3c47e',
      base: '#5a2e14',
      inlay: '#d9a84e',
      rimKind: 'teak',
      grain: '#4a2a10',
      edge: 'bullnose',
      pedestal: 'capstan',
    },
    accent: '#5fc8ff',
    accentDeep: '#153f5e',
    weathers: ['clear', 'cloudy', 'rain', 'storm', 'fog'],
  },
}

export const MAP_IDS = Object.keys(MAPS) as MapId[]

/**
 * Resolves a wire `map_id` to its definition, or null.
 *
 * Null is a first-class answer, not a failure: a lobby has no map yet, and a
 * server that ships a new one before the client has its scene must degrade to
 * the built-in felt rather than to a blank table.
 */
export function resolveMap(id: string | null | undefined): MapDef | null {
  if (!id) return null
  return MAPS[id as MapId] ?? null
}

/** A room at an hour under a sky: everything the renderer is handed. */
export interface SceneSpec {
  map: MapDef
  time: TimeOfDay
  weather: Weather
}

/**
 * The three wire ids as one scene, or null for a map this client has no scene
 * for. The hour and the sky degrade one at a time rather than taking the scene
 * with them: a weather this build does not know, or one the map does not list,
 * is dealt clear, and an unknown hour is dealt in daylight. A reload must never
 * lose the room over a word.
 */
export function resolveScene(
  mapId: string | null | undefined,
  time: string | null | undefined,
  weather: string | null | undefined,
): SceneSpec | null {
  const map = resolveMap(mapId)
  if (!map) return null
  const t: TimeOfDay = isTime(time) ? time : 'day'
  const w: Weather = isWeather(weather) && map.weathers.includes(weather) ? weather : 'clear'
  return { map, time: t, weather: w }
}

/** One string per distinct scene: the render cache's key and the builders' seed. */
export function sceneKey(spec: SceneSpec): string {
  return `${spec.map.id}:${spec.time}:${spec.weather}`
}

export { TIMES, WEATHERS }
export type { TimeOfDay, Weather }

File: server/game/maps.go
package game

// Maps: the room a match is played in, and the moment it is played at.
//
// A map is a scene, a table, an accent colour and a name. It changes no rule,
// no card, no timing; the domain carries it purely so that the *choice* can be
// the server's.
//
// That choice has to be server-side even though the consequence is entirely
// visual. LOCO is built to be watched: two players in one room describing two
// different tables to a viewer is a table that does not exist, and a highlight
// clip cut between two seats would jump between two rooms. Deriving it from the
// room code client-side would agree just as well, but it would also mean a room
// keeps the same map forever, and a rematch is meant to feel like a new match.
//
// A match is dealt at an hour and under a sky as well as in a room, and both are
// drawn here for the same reason: the client renders the scene from the three
// ids, so every seat has to be handed the same three.
type MapID string

const (
	// MapNeon is a rooftop terrace above a neon city.
	MapNeon MapID = "neon"
	// MapRune is the square of a village around an arcane tavern.
	MapRune MapID = "rune"
	// MapVelvet is a boulevard of art-deco hotels and marquees.
	MapVelvet MapID = "velvet"
	// MapOrbit is a base on an airless moon, under the home planet.
	MapOrbit MapID = "orbit"
	// MapSakura is a hot-spring village under cherry trees.
	MapSakura MapID = "sakura"
	// MapMarina is a harbour front: pier, lighthouse, boats.
	MapMarina MapID = "marina"
)

// MapIDs is every map a match can be dealt into, and the list the draw reads.
//
// The client keeps its own registry of the same ids (the scene, the table's
// materials, the accent) and falls back to the built-in felt for an id it does
// not know, so adding a map here before the client ships its scene degrades to
// the plain table rather than to a blank screen. `maps.test.ts` pins the
// client's list to this one.
var MapIDs = []MapID{MapNeon, MapRune, MapVelvet, MapOrbit, MapSakura, MapMarina}

// TimeOfDay is the hour a match is dealt at. It decides the light the scene is
// rendered in and the tint the table's highlight takes, and nothing else.
type TimeOfDay string

const (
	TimeDawn  TimeOfDay = "dawn"
	TimeDay   TimeOfDay = "day"
	TimeDusk  TimeOfDay = "dusk"
	TimeNight TimeOfDay = "night"
)

// TimesOfDay is every hour a match can be dealt at, and the list the draw reads.
var TimesOfDay = []TimeOfDay{TimeDawn, TimeDay, TimeDusk, TimeNight}

// Weather is the sky a match is dealt under. Like the hour it is presentation
// only: rain on the roof changes nothing about a card.
type Weather string

const (
	WeatherClear  Weather = "clear"
	WeatherCloudy Weather = "cloudy"
	WeatherRain   Weather = "rain"
	WeatherStorm  Weather = "storm"
	WeatherSnow   Weather = "snow"
	WeatherFog    Weather = "fog"
)

// Weathers is every weather any map can be dealt under.
var Weathers = []Weather{WeatherClear, WeatherCloudy, WeatherRain, WeatherStorm, WeatherSnow, WeatherFog}

// MapWeathers is the weather each room can be dealt under, and the list its
// draw reads. A room says what its sky can do: it does not snow on a moon with
// no air, and a harbour is more often in fog than a rooftop above the clouds.
// Every list is non-empty and `clear` is on every one of them, which
// maps_test.go pins.
var MapWeathers = map[MapID][]Weather{
	MapNeon:   {WeatherClear, WeatherCloudy, WeatherRain, WeatherStorm, WeatherSnow, WeatherFog},
	MapRune:   {WeatherClear, WeatherCloudy, WeatherRain, WeatherStorm, WeatherSnow, WeatherFog},
	MapVelvet: {WeatherClear, WeatherCloudy, WeatherRain, WeatherSnow, WeatherFog},
	MapOrbit:  {WeatherClear, WeatherFog, WeatherStorm},
	MapSakura: {WeatherClear, WeatherCloudy, WeatherRain, WeatherSnow, WeatherFog},
	MapMarina: {WeatherClear, WeatherCloudy, WeatherRain, WeatherStorm, WeatherFog},
}

// Valid reports whether m is a map this server can deal into.
func (m MapID) Valid() bool {
	for _, id := range MapIDs {
		if id == m {
			return true
		}
	}
	return false
}

// Weathers is the skies this room can be dealt under. Nil for a map the server
// does not know.
func (m MapID) Weathers() []Weather {
	return MapWeathers[m]
}

// Valid reports whether t is an hour a match can be dealt at.
func (t TimeOfDay) Valid() bool {
	for _, v := range TimesOfDay {
		if v == t {
			return true
		}
	}
	return false
}

// Valid reports whether w is a weather any map can be dealt under.
func (w Weather) Valid() bool {
	for _, v := range Weathers {
		if v == w {
			return true
		}
	}
	return false
}

// pickMap draws the room for a match. Uses the room's own rng so a seeded room
// deals a reproducible map alongside its reproducible hands.
func (r *Room) pickMap() MapID {
	if len(MapIDs) == 0 {
		return ""
	}
	return MapIDs[r.rng.Intn(len(MapIDs))]
}

// pickTime draws the hour a match is dealt at, from the same source as the map.
func (r *Room) pickTime() TimeOfDay {
	if len(TimesOfDay) == 0 {
		return ""
	}
	return TimesOfDay[r.rng.Intn(len(TimesOfDay))]
}

// pickWeather draws the sky for a match, among the ones the room allows. A map
// with no list (which no shipped map has) is dealt clear rather than nothing:
// the client would render a clear sky for an empty id anyway, and an empty
// field is one more thing a reload has to explain.
func (r *Room) pickWeather(m MapID) Weather {
	list := m.Weathers()
	if len(list) == 0 {
		return WeatherClear
	}
	return list[r.rng.Intn(len(list))]
}
